CMakeLists.txt

```diff
@@ -76,8 +76,18 @@ else()
 endif()
 FetchContent_MakeAvailable(SwiftFoundationICU SwiftFoundation)
 
+include(CheckSymbolExists)
 include(CheckLinkerFlag)
 
+check_symbol_exists(strlcat "string.h" HAVE_STRLCAT)
+check_symbol_exists(strlcpy "string.h" HAVE_STRLCPY)
```

Comment:
Is there a way that we can replicate this behavior in the package manifest as well so that when building via SwiftPM we get the same behavior? Or a way that we can add this check into the code itself rather than on the CMake/SwiftPM side?

Reply:
Generally, no, SwiftPM doesn't have a way to add such checks or generate "config.h" files.

However, in these cases (HAVE_STRLCAT and HAVE_STRLCPY in Swift), I think we can change swift-corelibs-foundation (and other projects) to use local versions of these functions only if certain conditions are not met. Below is an example of how we can do this:

```c
#if !TARGET_OS_MAC
#if defined(__linux__) && ((defined(__GLIBC__) && __GLIBC__ >= 2 && defined(__GLIBC_MINOR__) && __GLIBC_MINOR__ >= 38) || !defined(__GLIBC__))
// strlcat and strlcpy are available in the system
#else
// Define local versions of strlcpy and strlcat for glibc < 2.38
CF_INLINE size_t
strlcpy(char * dst, const char * src, size_t maxlen) {
  ...
}
#endif
#endif  // !TARGET_OS_MAC
```

This code assumes the following things:
- Swift is supported on macOS (and other Apple platforms), Linux, and Windows.
- Android's bionic supports strlcpy and strlcat. [Link](https://cs.android.com/android/platform/superproject/main/+/main:bionic/libc/include/string.h;drc=c5d9795cf403febea9562c03ffdb397327189ab0;l=96)
- Other than Android, Linux uses glibc or musl. Musl supports strlcpy and strlcat. [Link](https://git.musl-libc.org/cgit/musl/tree/include/string.h?id=ab31e9d6a0fa7c5c408856c89df2dfb12c344039#n86)
- glibc 2.38 and later supports strlcpy and strlcat.
- Windows doesn't have strlcpy and strlcat.

This approach is somewhat error-prone and may not be the best way to handle this, but I think this is the only way to handle this in SwiftPM. Also, I don't think supported platforms change frequently, so this approach should be fine with appropriate compile flags. Do you think this approach is acceptable?

Reply:
I think this is an alright way to right this given that SwiftPM can't perform a check - assuming there's no other macro defined in the header that we can use to indicate the presence or lack of this function. @parkera had to do something similar for CURL: https://github.com/apple/swift-corelibs-foundation/blob/15ee5c32d2b808fd912c586c535feaf9af2eb2d5/Sources/_CFURLSessionInterface/include/CFURLSessionInterface.h#L39-L57

The only tweak I'd suggest - rather than doing this for `!TARGET_OS_MAC`, should we just only do this for `TARGET_OS_LINUX` (or whatever the right macro is for that)? In other words, do we need to define `strlcpy` for Windows/Android or does the source code in Foundation already handle not calling that function?

Reply:
If you think it is a good idea to remove autoconf-style feature test preprocessor macros and check features in the code for compatibility with SwiftPM, I'll revisit this patch to do so.

Reply:
👍 yep since the alternative would break the SwiftPM build of this project, checking conditionally in the code is the way to go here to avoid breaking the build